game-core/src/test/java/games/strategy/triplea/delegate/power/calculator/TotalPowerAndTotalRollsTest.java

```diff
@@ -1,33 +1,723 @@
+import static games.strategy.triplea.delegate.battle.BattleState.Side.OFFENSE;
+import static games.strategy.triplea.delegate.battle.steps.MockGameData.givenGameData;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.is;
-import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
 import games.strategy.engine.data.GameData;
 import games.strategy.engine.data.GamePlayer;
+import games.strategy.engine.data.MutableProperty;
 import games.strategy.engine.data.Territory;
+import games.strategy.engine.data.TerritoryEffect;
 import games.strategy.engine.data.Unit;
 import games.strategy.engine.data.UnitType;
+import games.strategy.triplea.attachments.RulesAttachment;
+import games.strategy.triplea.attachments.TerritoryEffectAttachment;
 import games.strategy.triplea.attachments.UnitAttachment;
+import games.strategy.triplea.attachments.UnitSupportAttachment;
+import games.strategy.triplea.delegate.Die;
 import games.strategy.triplea.delegate.GameDataTestUtil;
+import games.strategy.triplea.delegate.battle.BattleState;
 import games.strategy.triplea.xml.TestMapGameData;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
 import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
 import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInstance;
 import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
 import org.mockito.Mock;
 import org.mockito.junit.jupiter.MockitoExtension;
+import org.triplea.java.collections.IntegerMap;
+import org.triplea.util.Triple;
+import org.triplea.util.Tuple;
 
+@ExtendWith(MockitoExtension.class)
 class TotalPowerAndTotalRollsTest {
 
+  enum BattleType {
+    NORMAL,
+    AA,
+  }
+
+  @Mock GamePlayer owner;
+
+  private Unit givenUnit(
+      final String name,
+      final BattleType type,
+      final GameData gameData,
+      final BattleState.Side side,
+      final int power,
+      final int roll)
+      throws MutableProperty.InvalidValueException {
+    return givenUnit(givenUnitType(name, gameData), type, side, power, roll);
+  }
+
+  private Unit givenUnit(
+      final UnitType unitType,
+      final BattleType type,
+      final BattleState.Side side,
+      final int power,
+      final int roll)
+      throws MutableProperty.InvalidValueException {
+    final Unit unit = unitType.create(1, owner, true).get(0);
+    if (type == BattleType.NORMAL) {
```

Comment:
The logic in the test setup code jumps out at me. I'd look for ways to avoid this and make the test setup more direct.

It looks like we might be able to do that by using a map. WDYT?

For example:

````
givenUnit(UnitType, Map<String, Integer>)
 ....
 
)

Usage:
 Map.of(UnitAttachment.DEFENSE_STRENGTH, power, UnitAttachment.DEFENSE_ROLL, roll);

```

Reply:
In that case, I would just go back to how I had it before:
```
Unit unit = givenUnit(unitType);
unit.getUnitAttachment().getPropertyOrThrow(DEFENSE_STRENGTH).setValue(power);
unit.getUnitAttachment().getPropertyOrThrow(DEFENSE_ROLL).setValue(power);
```

I switched to the current way because it made the tests more concise and easier to read (at least for me).  But if you want it more verbose, I can go back.

Reply:
That does seem kinda clunky. 

```
Unit unit = givenUnit(unitType, Map.of(DEFENSE_STRENGTH, power, DEFENSE_ROLL, power);
```

```
Unit unit = givenUnit(UnitParameters.builder()
    .type(unitType)
    .defenseStrength(power)
     .defenseRolls(power)
     .build());
```

Looking at the source code that we are using, notably:
```
unit.getUnitAttachment().getPropertyOrThrow(DEFENSE_STRENGTH).setValue(power);
```
That is already very clunky (and a violation of the law of demeter). Seemingly it could be as simple as:
```
unit.setDefenseStrength(power);
```
The above has encapsulation, and notably hides the magic string property values. We may want to really think about going to that level of change before jumping there. A 4th alternative would be to add the setter method to unit attachment:

```
unit.getUnitAttachment().setProperty(DEFENSE_STRENGTH, power);
```
And that is still clunky (notably the 'orThrow' is there because we have a string parameter), so a 5th option is:

```
unit.getUnitAttachment().setDefenseStrength(power);
```

> But if you want it more verbose

It's not verbosity, two goals:
- test being self contained and understandable as much as possible by reading just the single test
- avoiding logic in the tests (arguably that should also be tested!)

Reply:
UnitAttachment already has setters.  They are all private, for some reason.  I've made a few public and marked them with `@VisibleForTesting` and also changed them to return `this` so that I can call them in a fluid manner.  I've updated the tests to use the setters instead of `getPropertyOrThrow().setValue`.